darshan-runtime/lib/darshan-pnetcdf-api.m4

```diff
@@ -0,0 +1,760 @@
+            }
+            DARSHAN_TIMER_INC_NO_OVERLAP(
+                rec_ref->var_rec->fcounters[PNETCDF_VAR_F_READ_TIME],
+                tm1, tm2, rec_ref->last_read_end);',
+`$1',`put',`rec_ref->var_rec->counters[PNETCDF_VAR_BYTES_WRITTEN] += $3;
+            rec_ref->var_rec->counters[`PNETCDF_VAR_PUT_VAR'Upcase($2)] += 1;
+            if (rec_ref->last_io_type == DARSHAN_IO_READ)
+                rec_ref->var_rec->counters[PNETCDF_VAR_RW_SWITCHES] += 1;
+            rec_ref->last_io_type = DARSHAN_IO_WRITE;
+            DARSHAN_BUCKET_INC(
+                &(rec_ref->var_rec->counters[PNETCDF_VAR_SIZE_WRITE_AGG_0_100]), $3);
+            common_access_vals[0] = $3;
+            cvc = darshan_track_common_val_counters(&rec_ref->access_root,
+                common_access_vals, PNETCDF_VAR_MAX_NDIMS+PNETCDF_VAR_MAX_NDIMS+1, &rec_ref->access_count);
+            if (cvc) DARSHAN_UPDATE_COMMON_VAL_COUNTERS(
+                &(rec_ref->var_rec->counters[PNETCDF_VAR_ACCESS1_ACCESS]),
+                &(rec_ref->var_rec->counters[PNETCDF_VAR_ACCESS1_COUNT]),
+                cvc->vals, cvc->nvals, cvc->freq, 0);
+            if (rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_START_TIMESTAMP] == 0 ||
+             rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_START_TIMESTAMP] > tm1)
+                rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_START_TIMESTAMP] = tm1;
+            rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_END_TIMESTAMP] = tm2;
+            if (rec_ref->var_rec->fcounters[PNETCDF_VAR_F_MAX_WRITE_TIME] < tm2 - tm1) {
+                rec_ref->var_rec->fcounters[PNETCDF_VAR_F_MAX_WRITE_TIME] = tm2 - tm1;
+                rec_ref->var_rec->counters[PNETCDF_VAR_MAX_WRITE_TIME_SIZE] = $3;
+            }
+            DARSHAN_TIMER_INC_NO_OVERLAP(
+                rec_ref->var_rec->fcounters[PNETCDF_VAR_F_WRITE_TIME],
+                tm1, tm2, rec_ref->last_write_end);')')dnl
+dnl
+dnl
+define(`UPDATE_INDEPCOLL_RW_COUNTER',`ifelse(
+       `$1$2',`get',`rec_ref->var_rec->counters[PNETCDF_VAR_INDEP_READS] += 1',
+       `$1$2',`get_all',`rec_ref->var_rec->counters[PNETCDF_VAR_COLL_READS] += 1',
+       `$1$2',`put',`rec_ref->var_rec->counters[PNETCDF_VAR_INDEP_WRITES] += 1',
+       `$1$2',`put_all',`rec_ref->var_rec->counters[PNETCDF_VAR_COLL_WRITES] += 1')')dnl
+dnl
+
+dnl
+define(`APINAME',`ifelse(`$3',`',`ncmpi_$1_var$2$4',`ncmpi_$1_var$2$3$4')')dnl
+dnl
+
+/* ncmpi_get/put_var<kind>_<type>_<mode> API */
+dnl
+dnl GETPUT_API(get/put, `'/1/a/s/m, `'/itype, `'/_all)
+dnl
+define(`GETPUT_API',`
+DARSHAN_FORWARD_DECL(APINAME($1,$2,$3,$4), int, (int ncid, int varid, ArgKind($2) BufArgs($1,$3)));
+
+int DARSHAN_DECL(APINAME($1,$2,$3,$4))(int ncid, int varid, ArgKind($2)BufArgs($1,$3))
+{
+    int err, ret;
+    MPI_Offset $1_before, $1_after;
+    double tm1, tm2;
+
+    MAP_OR_FAIL(APINAME($1,$2,$3,$4));
+
+    err = ncmpi_inq_$1_size(ncid, &$1_before);
+    if (err != NC_NOERR) return err;
```

Comment:
The only chance for ncmpi_inq_put/get_size to fail is the file ID ncid or buffer 1_before is invalid. Invalid ncid should be detected long before.
Invalid buffer is not possible here as 1_before is defined in line 140.

Reply:
This call occurs before the underlying I/O call, so presumably PnetCDF has not had a chance to sanity check a bogus `ncid` value from the app. In any case, it's a pretty low stakes change to just `return(ret)` instead of `return(err)` here, so that in the unlikely event they do fail, the I/O call return code is actually passed back to the user. Probably best practice in case the logic were copied somewhere else, too.

Reply:
If ncid is bogus, then the error returned by ncmpi_inq_$1_size will be the same as one returned by ncmpi_get/put_varxxx, because they both run the same sanity check for ncid.

Reply:
I understand what @wkliao is saying, but I'm with @shanedsnyder on this.  Darshan tries extremely hard not to introduce additional failure points when it can be helped (even if unlikely) because the stakes are high in terms of user/facility trust for Darshan deployment.  The assumptions about the safety of the call could be broken if surrounding Darshan code or code within the PnetCDF library were to change at some point in the future.

Sometimes critical error paths can't be avoided, but it seems pretty easy to fail gracefully here.

Reply:
Are you suggesting to skip the error checking of `ncmpi_inq_$1_size()` ?

Reply:
The key thing is that Darshan wrappers should _always_ invoke the underlying I/O call, and they should _always_ return the error code as given by the I/O call. As Phil mentions, it's the absolute safest way to ensure our wrappers aren't affecting application I/O behavior, even if we think it's okay in this particular code path right now.

Since these calls are invoked twice (once before and once after the I/O call), I think what we could do is just skip error checking on the first call (before I/O routine) so that there is no possibility of the I/O call being skipped. We can then error check the 2nd call, and if it returns an error, then we just return the error code of the I/O routine (i.e., return `ret`, rather than `err`) and skip all Darshan instrumentation.

Reply:
I fixed the last remaining instance of this in [4039ca6](https://github.com/darshan-hpc/darshan/pull/456/commits/4039ca6c5303d004ade115c5f58f041921749731) (the other calls were removed as a side effect of [2edfbc7](https://github.com/darshan-hpc/darshan/pull/456/commits/2edfbc71d7f52eff0021915be6d30e9a48ad7d39)).

If somehow `ncmpi_inq_get/put_size()` fails, then the file-level `PNETCDF_FILE_BYTES_READ/WRITTEN` counters are set to -1 to mark them as invalid. In any case, the return value from Darshan's wrapper is now _always_ the return value of the underlying `ncmpi_close()` call.